op-deployer/pkg/deployer/verify/etherscan.go

```diff
@@ -0,0 +1,263 @@
+package verify
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"golang.org/x/time/rate"
+)
+
+type EtherscanGenericResp struct {
+	Status  string `json:"status"`
+	Message string `json:"message"`
+	Result  string `json:"result"`
+}
+
+type EtherscanContractCreationResp struct {
+	Status  string `json:"status"`
+	Message string `json:"message"`
+	Result  []struct {
+		ContractCreator string `json:"contractCreator"`
+		TxHash          string `json:"txHash"`
+	} `json:"result"`
+}
+
+type EtherscanClient struct {
+	apiKey      string
+	url         string
+	rateLimiter *rate.Limiter
+}
+
+func getAPIEndpoint(l1ChainID uint64) string {
+	switch l1ChainID {
+	case 1:
+		return "https://api.etherscan.io/api" // mainnet
+	case 11155111:
+		return "https://api-sepolia.etherscan.io/api" // sepolia
+	default:
+		return ""
+	}
+}
+
+func NewEtherscanClient(apiKey string, url string, rateLimiter *rate.Limiter) *EtherscanClient {
```

Comment:
would love to see some tests for this function. you can use a mock http server to help here.

Reply:
Done in https://github.com/ethereum-optimism/optimism/pull/14633/commits/f96024b99ce03d188cf2a22eb7e9dc38f9b1e313